### PR TITLE
Driver structure refactor

### PR DESCRIPTION
**Pull request by hellkite500**
Addressing some long standing technical debt, this PR aims to simplify the calling/driving code in `main`, and implements/extends HY_Features concepts to achieve this.

## Additions

- `Network` class responsible for keeping the topology of the hydrologic network as a lightweight, efficient index.
- `HY_Features` a high level hydrologic feature collection, creates and maintains holistic catchments (`HY_Catchment`) and related objects.

## Removals

- Significant driver hacks built up over time.

## Changes

- Implement all formulations as `Catchment_Formulation`.  Move catchment output handling to catchment output streams.

## Testing

1. New tests for network and hy_features classes.

## Todos

- Additional unit tests additions to HY_Features class.

## Checklist

- [x] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

- [x] Linux
- [x] MacOS

### CODE REVIEW

File: include/utilities/FileStreamHandler.hpp [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
 namespace utils
 {
-    class FileStreamHandler() : public StreamHandler
+    class FileStreamHandler : public StreamHandler
     {
-            FileStreamHandler(const char*) : StreamHandler()
+      public:
+            FileStreamHandler(const char* path) : StreamHandler()
             {
-                output_stream = make_shared(new(std::fstream(path)));
```
Comment (donaldwj) [inline include/utilities/FileStreamHandler.hpp:12]: Is it possible that this change is why you where having issues with fstream objects>

**Review by donaldwj — APPROVED** [review]
I don't see anything obvious out of place. I will want to look over the test output.

File: include/utilities/FileStreamHandler.hpp (hunk 1 above)
Comment (hellkite500) [reply include/utilities/FileStreamHandler.hpp:12]: This wouldn't compile.

File: include/core/HY_Features.hpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      private:
+        std::unordered_map<std::string, std::shared_ptr<HY_Catchment>> _catchments;
+        std::unordered_map<std::string, std::shared_ptr<HY_HydroNexus>> _nexuses;
+        network::Network network;
+        std::shared_ptr<Formulation_Manager> formulations;
+
+    };
+}
+
+#endif //HY_GRAPH_H
```
Comment (christophertubbs) [inline include/core/HY_Features.hpp:85]: The comment after the endif has the wrong name, but that's not an issue that will probably ever affect anything; just something I noticed.

File: src/core/HY_Features.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      std::string feat_id;
+      std::string feat_type;
+      std::vector<std::string> origins, destinations;
+
+      for(const auto& feat_idx : network){
+        feat_id = network.get_id(feat_idx);//feature->get_id();
+        feat_type = feat_id.substr(0, 3);
+
+        destinations  = network.get_destination_ids(feat_id);
+        if(feat_type == "cat")
```
Comment (christophertubbs) [inline src/core/HY_Features.cpp:27]: I wish we had a way to get away from having different logic based on the first three characters of an ID, but that a discussion for another day.

**Review by christophertubbs — APPROVED** [review]
I think these are all really good changes. I like the direction this is all going in. I added a few comments, but those are all just mild thoughts.

**Pull request merged**